base/regex.jl

```diff
@@ -542,12 +539,59 @@ r"julia"
 ```
 """
 function *(r1::Union{Regex,AbstractString,AbstractChar}, rs::Union{Regex,AbstractString,AbstractChar}...)
-    opts = unique((r.compile_options, r.match_options) for r in (r1, rs...) if r isa Regex)
-    length(opts) == 1 ||
-        throw(ArgumentError("cannot multiply regexes with incompatible options"))
-    Regex(string(unwrap_string(r1), unwrap_string.(rs)...), opts[1][1], opts[1][2])
+    mask = PCRE.CASELESS | PCRE.MULTILINE | PCRE.DOTALL | PCRE.EXTENDED # imsx
+    match_opts   = typemax(UInt32) # all args must agree on this
+    compile_opts = typemax(UInt32) # all args must agree on this
+    shared = mask
+    for r in (r1, rs...)
+        r isa Regex || continue
+        if match_opts == typemax(UInt32)
+            match_opts = r.match_options
+            compile_opts = r.compile_options & ~mask
+        else
+            r.match_options == match_opts &&
+                r.compile_options & ~mask == compile_opts ||
+                throw(ArgumentError("cannot multiply regexes: incompatible options"))
+        end
+        shared &= r.compile_options & mask
+    end
+    unshared = mask & ~shared
+    Regex(string(unwrap_string(r1, unshared), unwrap_string.(rs, Ref(unshared))...), compile_opts | shared, match_opts)
+end
+
+unwrap_string(r::Regex, unshared::UInt32) = string("(?", regex_opts_str(r.compile_options & unshared), ':', r.pattern, ')')
+unwrap_string(s::Union{AbstractString,AbstractChar}, ::UInt32) = string("(?:", s, ')')
```

Comment:
So I think this is not how we want this to work: I think that a string or character should be converted to a regex that only matches that exact string or character, not to a regex with that string or character as its content. Also, I think there's a case to be made for making this a method of convert. For a clearer example where it matters, should `r"a|b" * "c|d"` produce `r"(?:a|b)(?:c|d)"` which would match `ac`, `ad`, `bc` and `bd` or should it produce something like `r"(?:a|b)(?:c\|d)"` which would match `ac|d` and `bc|d`? I think it has to be the latter.

Reply:
OK, there seems to be a [solution](https://www.pcre.org/original/doc/html/pcrepattern.html#SEC5) after all: enclose the string between "\Q" and "\E". Remains to decide which behavior is the best, I don't have a strong opinion. But it's true that if one wants the regex interpretation of characters, it's easy enough to wrap the string in a call to `Regex`.